client/client.go

```diff
@@ -89,4 +89,29 @@ type OpAMPClient interface {
 	// May be called before or after Start().
 	// May be also called from OnMessage handler.
 	RequestConnectionSettings(request *protobufs.ConnectionSettingsRequest) error
+
+	// SetCustomCapabilities modifies the set of customCapabilities supported by the client.
+	// The new customCapabilities will be sent with the next message to the server.
+	// May be called anytime after Start(), including from OnMessage handler.
+	// nil values are not allowed and will return an error.
+	//
+	// See
+	// https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#customcapabilities
+	// for more details.
+	SetCustomCapabilities(customCapabilities *protobufs.CustomCapabilities) error
```

Comment:
Do we need to be able to change custom capabilities after Start? For standard capabilities we allow specifying them in StartSettings only and can't change after Start. Any reason not to do that for custom one's as well?

Reply:
It's easy to allow them to be changed and I think it's useful to support it. Agents may support changing configuration or adding packages that modify (enable or disable) custom capabilities.

Reply:
I am worried about introducing races here. Let's say client removes a capability but before it gets delivered to the server the server sends a custom message of that capability. Do we expect client to ignore that? What about the opposite situation? We add a capability but before it is delivered a message from the server arrives but the client ignores it.

This may end up resulting in surprises and I don't think we have a good way to prevent them. There is just no way for the client to know the changed capabilities set is delivered and processed by the server so the client has to always assume the server may send custom messages that don't match what the client expects.

Reply:
I appreciate your concerns.

> Let's say client removes a capability but before it gets delivered to the server the server sends a custom message of that capability.

The client will ignore any messages that are not supported by the client. If the server is not yet aware that a message is no longer supported, it may send still send messages, but they will be ignored. If a server needs to know if a message was successfully received and processed, the custom capability should be implemented to send a response and the server will receive confirmation that the message was received.

> What about the opposite situation? We add a capability but before it is delivered a message from the server arrives but the client ignores it.

As soon as the client sets the message, the client will receive messages with that capability. This is because SetCustomCapabilities modifies ClientSyncedState and HasCustomCapability uses that field to determine if the message should be delivered. That said, I think that messages with the capability would only be sent once the server has received the updated list of supported capabilities.

I see two primary patterns for custom messages: streaming messages and request/response messages. It is up to the implementer to choose and implement the appropriate pattern.

For streaming messages (server to agent or agent to server), the only risk is that the streaming is delayed until the capability is known to be supported or that some messages are ignored until the capability is known to not be supported.

For request/response style messages, the sender should expect a response to its message and can handle that as appropriate.

Reply:
@tigrannajaryan I removed CustomCapabilities from the client StartSettings.